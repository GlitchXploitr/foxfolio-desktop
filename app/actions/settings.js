// @flow
import type { SettingsType } from '../reducers/settings';
import type { Action } from './action.d';

export function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}
