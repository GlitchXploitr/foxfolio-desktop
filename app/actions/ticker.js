// @flow
import * as R from 'ramda';
import type { Exchanges } from '../reducers/exchanges/types.d';
import type { Action, Dispatch, GetState, ThunkAction } from './action.d';
import startTimer from './timer';
import type { Wallet } from '../reducers/wallets/types.d';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(ticker: Object): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker,
  };
}

function receiveHistory(fsym: string, tsym: string, history: [{ close: number }]): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Object): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [], fiatCurrency?: string, cryptoCurrency?: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency || state.settings.fiatCurrency,
      cryptoCurrency || state.settings.cryptoCurrency,
      extraSymbols);
    const fsyms = symbols.from.join(',');
    const tsyms = symbols.to.join(',');
    if (fsyms && tsyms) {
      fetch(`https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`)
        .then(result => result.json())
        .then(result => dispatch(receiveTickerUpdate(result.RAW)))
        .catch(error => console.error(error));
    }
  };
}

export function requestHistory(fsym: string, tsym: string, forceRequest: boolean = false): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = new Date((new Date()).getTime() - (1000 * 60 * 2));
    if (forceRequest
      || !(getState().ticker.history[fsym]
        && getState().ticker.history[fsym][tsym]
        && getState().ticker.history[fsym][tsym].lastUpdate >= twoMinutesAgo)
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker() {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timer.ticker) {
      const timer = setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]): { from: string[], to: string[] } {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = R.values(exchanges)
    .map(exchange => exchange.balances)
    .reduce((acc, balances) => acc.concat(R.keys(balances)), []);

  return R.map(R.uniq)(
    {
      from: [cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols],
      to: [cryptoCurrency, fiatCurrency],
    });
}
