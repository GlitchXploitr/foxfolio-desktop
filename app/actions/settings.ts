import { SettingsType } from '../reducers/settings';
import { Action, Dispatch, ThunkAction } from './actions.types';
import { requestTickerUpdate } from './ticker';

function saveSettings(settings: SettingsType): Action {
  return {
    type: 'SAVE_SETTINGS',
    settings,
  };
}

export function saveSettingsAndUpdateTicker(settings: SettingsType): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(saveSettings(settings));
    dispatch(requestTickerUpdate([], settings.fiatCurrency, settings.cryptoCurrency));
  };
}
