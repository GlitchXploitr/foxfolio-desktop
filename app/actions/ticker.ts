import * as R from 'ramda';
import { Coinlist } from '../reducers/coinlist';
import { Exchanges } from '../reducers/exchanges.types';
import { Ticker } from '../reducers/ticker';
import { Wallet } from '../reducers/wallets';
import { Action, Dispatch, GetState, ThunkAction } from './actions.types';
import startTimer from './timer';

const REFRESH_TIME_IN_MS = 30000;

function fetchingTickerUpdate(): Action {
  return {
    type: 'LAST_UPDATED',
    key: 'ticker',
    time: new Date(),
  };
}

function receiveTickerUpdate(ticker: Ticker): Action {
  return {
    type: 'TICKER_UPDATE',
    ticker,
  };
}

function receiveHistory(fsym: string, tsym: string, history: Array<{ close: number }>): Action {
  return {
    type: 'HISTORY_UPDATE',
    fsym,
    tsym,
    history,
  };
}

function receiveCoinList(coinlist: Coinlist): Action {
  return {
    type: 'RECEIVE_COIN_LIST',
    coinlist,
  };
}

export function requestTickerUpdate(
  extraSymbols: string[] = [],
  fiatCurrency?: string,
  cryptoCurrency?: string
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchingTickerUpdate());

    const state = getState();
    const symbols = getSymbolsFromTransactions(
      state.exchanges,
      state.wallets,
      fiatCurrency || state.settings.fiatCurrency,
      cryptoCurrency || state.settings.cryptoCurrency,
      extraSymbols
    );
    const fsyms = symbols.from.join(',');
    const tsyms = symbols.to.join(',');
    if (fsyms && tsyms) {
      fetch(`https://min-api.cryptocompare.com/data/pricemultifull?fsyms=${fsyms}&tsyms=${tsyms}`)
        .then(result => result.json())
        .then(result => dispatch(receiveTickerUpdate(result.RAW)))
        .catch(error => console.error(error));
    }
  };
}

export function requestHistory(
  fsym: string,
  tsym: string,
  forceRequest: boolean = false
): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const twoMinutesAgo = new Date(new Date().getTime() - 1000 * 60 * 2);
    if (
      forceRequest ||
      !(
        getState().ticker.history[fsym] &&
        getState().ticker.history[fsym][tsym] &&
        getState().ticker.history[fsym][tsym].lastUpdate >= twoMinutesAgo
      )
    ) {
      fetch(`https://min-api.cryptocompare.com/data/histominute?fsym=${fsym}&tsym=${tsym}`)
        .then(result => result.json())
        .then(result => dispatch(receiveHistory(fsym, tsym, result.Data)))
        .catch(error => console.error(error));
    }
  };
}

export function requestCoinList(): ThunkAction {
  return (dispatch: Dispatch) => {
    fetch('https://min-api.cryptocompare.com/data/all/coinlist')
      .then(result => result.json())
      .then(result => dispatch(receiveCoinList(result.Data)))
      .catch(error => console.error(error));
  };
}

export function continuouslyUpdateTicker(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    if (!getState().timers.timers.ticker) {
      const timer = window.setInterval(() => dispatch(requestTickerUpdate()), REFRESH_TIME_IN_MS);
      dispatch(startTimer('ticker', timer));
    }
  };
}

function getSymbolsFromTransactions(
  exchanges: Exchanges,
  wallets: Wallet[],
  fiatCurrency: string,
  cryptoCurrency: string,
  extraSymbols: string[]
): { from: string[]; to: string[] } {
  const walletSymbols = wallets.map(wallet => wallet.currency) || [];
  const exchangeSymbols = R.pipe(
    R.values,
    R.map(exchange => exchange.balances),
    R.map(R.keys),
    R.reduce((acc, keys) => acc.concat(keys), [] as string[])
  )(exchanges);

  return {
    from: R.uniq([cryptoCurrency, ...walletSymbols, ...exchangeSymbols, ...extraSymbols]),
    to: [cryptoCurrency, fiatCurrency],
  };
}
