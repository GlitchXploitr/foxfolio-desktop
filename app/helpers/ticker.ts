import { Ticker, TickerEntry } from '../reducers/ticker';

export const getTickerEntry = (ticker: Ticker, fsym: string, tsym: string): TickerEntry => {
  if (fsym !== tsym) {
    if (ticker) {
      const tickerBase = ticker[fsym];
      if (tickerBase != null) {
        const tickerEntry = tickerBase[tsym];
        if (tickerEntry != null && tickerEntry.PRICE && tickerEntry.CHANGEPCT24HOUR) {
          return tickerEntry;
        }
      }
    }
    console.log(`Couldn't get ticker entry for ${fsym}/${tsym}`);
    return {
      PRICE: 0,
      CHANGEPCT24HOUR: 0,
    };
  } else {
    return { PRICE: 1, CHANGEPCT24HOUR: 0 };
  }
};
