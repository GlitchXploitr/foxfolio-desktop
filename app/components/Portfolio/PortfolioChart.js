// @flow
import React from 'react';
import { defaults, HorizontalBar } from 'react-chartjs-2';
import { withTheme } from 'material-ui';
import { getTickerPrice } from '../../helpers/transactions';
import getColor from '../../utils/colors';

// Disable animating charts by default.
defaults.global.animation = false;

type Props = {
  ticker: Object,
  portfolio: Object,
  sum: number,
  theme: Object
};

function PortfolioChart({ ticker, portfolio, sum, theme }: Props) {
  const chartData = calculateChartData(ticker, portfolio, sum);

  return (
    <HorizontalBar
      data={chartData}
      options={{
        responsive: true,
        maintainAspectRatio: false,
        tooltips: {
          mode: 'x',
          callbacks: {
            label: (item, data) => `${data.datasets[item.datasetIndex].label}: ${item.xLabel} %`,
          },
        },
        legend: {
          display: true,
          labels: {
            fontColor: theme.palette.text.secondary,
          },
        },
        scales: {
          xAxes: [
            {
              stacked: true,
              fontColor: theme.palette.text.secondary,
              ticks: {
                min: 0,
                max: 100,
              },
            }],
          yAxes: [
            {
              stacked: true,
              fontColor: theme.palette.text.secondary,
            }],
        },
      }}
    />
  );
}

export default withTheme()(PortfolioChart);

function calculateChartData(ticker: Object, portfolio: Object, sum: number) {
  const datasets = Object.keys(portfolio)
    .filter(asset => portfolio[asset] > 0)
    .filter(asset => ticker[asset])
    .sort((a, b) =>
      (getTickerPrice(ticker, b, 'BTC') * portfolio[b]) - (getTickerPrice(ticker, a, 'BTC') * portfolio[a]))
    .map(asset => ({
      label: asset.toUpperCase(),
      backgroundColor: getColor(asset),
      borderColor: '#fff',
      borderWidth: 1,
      data: [((getTickerPrice(ticker, asset, 'BTC') * portfolio[asset] * 100) / sum).toFixed(2)],
    }));

  return { datasets };
}
