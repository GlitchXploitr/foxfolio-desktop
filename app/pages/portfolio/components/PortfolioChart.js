// @flow
import React from 'react';
import { defaults, HorizontalBar } from 'react-chartjs-2';
import { withTheme } from 'material-ui';
import { getTickerPrice } from '../../../helpers/transactions';
import type { SettingsType } from '../../../reducers/settings';
import type { Ticker } from '../../../reducers/ticker/types.d';
import getColor from '../../../utils/colors';

// Disable animating charts by default.
defaults.global.animation = false;

type Props = {
  ticker: Ticker,
  portfolio: Object,
  settings: SettingsType,
  sum: number,
  theme: Object
};

function PortfolioChart({ ticker, portfolio, settings, sum, theme }: Props) {
  const chartData = calculateChartData(ticker, portfolio, settings, sum, theme);

  return (
    <HorizontalBar
      data={chartData}
      options={{
        responsive: true,
        maintainAspectRatio: false,
        tooltips: {
          mode: 'x',
          callbacks: {
            label: (item, data) => `${data.datasets[item.datasetIndex].label}: ${item.xLabel} %`,
          },
        },
        legend: {
          display: true,
          labels: {
            fontColor: theme.palette.text.primary,
          },
        },
        scales: {
          xAxes: [
            {
              stacked: true,
              gridLines: {
                display: false,
                color: theme.palette.text.secondary,
              },
              ticks: {
                fontColor: theme.palette.text.primary,
                min: 0,
                max: 100,
              },
            }],
          yAxes: [
            {
              stacked: true,
              gridLines: {
                color: theme.palette.text.secondary,
              },
            }],
        },
      }}
    />
  );
}

export default withTheme()(PortfolioChart);

function calculateChartData(ticker: Ticker, portfolio: Object, settings: SettingsType, sum: number, theme: Object) {
  const { cryptoCurrency } = settings;

  const datasets = Object.keys(portfolio)
    .filter(asset => portfolio[asset] > 0)
    .filter(asset => ticker[asset])
    .sort((a, b) =>
      (getTickerPrice(ticker, b, cryptoCurrency) * portfolio[b])
      - (getTickerPrice(ticker, a, cryptoCurrency) * portfolio[a]))
    .map(asset => ({
      label: asset.toUpperCase(),
      backgroundColor: getColor(asset),
      borderColor: theme.palette.background.default,
      borderWidth: 1,
      data: [((getTickerPrice(ticker, asset, cryptoCurrency) * portfolio[asset] * 100) / sum).toFixed(2)],
    }));

  return { datasets };
}
