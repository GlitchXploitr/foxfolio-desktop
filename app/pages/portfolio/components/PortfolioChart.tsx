import chartjs, { ChartData, ChartTooltipItem } from 'chart.js';
import { withTheme } from 'material-ui';
import { Theme } from 'material-ui/styles';
import React from 'react';
import { defaults, HorizontalBar } from 'react-chartjs-2';
import { getTickerPrice } from '../../../helpers/ticker';
import { SettingsType } from '../../../reducers/settings';
import { Ticker } from '../../../reducers/ticker';
import getColor from '../../../utils/colors';

// Disable animating charts by default.
(defaults as any).global.animation = false;

interface Props {
  ticker: Ticker;
  portfolio: { [key: string]: number };
  settings: SettingsType;
  sum: number;
}

export const PortfolioChart = withTheme()<Props>(({ ticker, portfolio, settings, sum, theme }) => {
  const chartData = calculateChartData(ticker, portfolio, settings, sum, theme);

  return (
    <HorizontalBar
      data={chartData}
      options={{
        responsive: true,
        maintainAspectRatio: false,
        tooltips: {
          mode: 'x',
          callbacks: {
            label: labelCallback,
          },
        },
        legend: {
          display: true,
          labels: {
            fontColor: theme.palette.text.primary,
          },
        },
        scales: {
          xAxes: [
            {
              stacked: true,
              gridLines: {
                display: false,
                color: theme.palette.text.secondary,
              },
              ticks: {
                fontColor: theme.palette.text.primary,
                min: 0,
                max: 100,
              },
            },
          ],
          yAxes: [
            {
              stacked: true,
              gridLines: {
                color: theme.palette.text.secondary,
              },
            },
          ],
        },
      }}
    />
  );
});

function calculateChartData(
  ticker: Ticker,
  portfolio: { [key: string]: number },
  settings: SettingsType,
  sum: number,
  theme: Theme
): chartjs.ChartData {
  const { cryptoCurrency } = settings;

  const datasets = Object.keys(portfolio)
    .filter(asset => portfolio[asset] > 0)
    .filter(asset => ticker[asset])
    .sort(
      (a, b) =>
        getTickerPrice(ticker, b, cryptoCurrency) * portfolio[b] -
        getTickerPrice(ticker, a, cryptoCurrency) * portfolio[a]
    )
    .map(asset => ({
      label: asset.toUpperCase(),
      backgroundColor: getColor(asset),
      borderColor: theme.palette.background.default,
      borderWidth: 1,
      data: [
        {
          x: (getTickerPrice(ticker, asset, cryptoCurrency) * portfolio[asset] * 100 / sum).toFixed(
            2
          ),
        },
      ],
    }));

  return { datasets };
}

const labelCallback = (item: ChartTooltipItem, data: ChartData) => {
  if (data && data.datasets && item && item.datasetIndex != null) {
    return `${data.datasets[item.datasetIndex].label}: ${item.xLabel} %`;
  }
  return '';
};
