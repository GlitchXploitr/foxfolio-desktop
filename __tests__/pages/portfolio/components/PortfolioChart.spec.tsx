import { createShallow } from '@material-ui/core/test-utils';
import * as React from 'react';
import { PortfolioChart } from '../../../../app/pages/portfolio/components/PortfolioChart';
import { initialSettings } from '../../../../app/reducers/settings';
import { Ticker } from '../../../../app/reducers/ticker';

let shallow;

beforeAll(() => {
  shallow = createShallow();
});

test('Render portfolio chart', () => {
  const portfolio = { BTC: 5, ETH: 10 };
  const ticker: Ticker = {
    BTC: { BTC: { PRICE: 1, CHANGEPCT24HOUR: 0 } },
    ETH: { BTC: { PRICE: 0.1, CHANGEPCT24HOUR: 2 } },
  };

  const wrapper = shallow(
    <PortfolioChart portfolio={portfolio} settings={initialSettings} sum={6} ticker={ticker} />
  ).dive();
  expect(wrapper).toMatchSnapshot();
});
