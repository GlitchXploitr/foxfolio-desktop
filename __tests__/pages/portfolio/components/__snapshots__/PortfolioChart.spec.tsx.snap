// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render portfolio chart 1`] = `
<HorizontalBar
  data={
    Object {
      "datasets": Array [
        Object {
          "backgroundColor": "#F7931A",
          "borderColor": "#fafafa",
          "borderWidth": 1,
          "data": Array [
            Object {
              "x": "83.33",
            },
          ],
          "label": "BTC",
        },
        Object {
          "backgroundColor": "#282828",
          "borderColor": "#fafafa",
          "borderWidth": 1,
          "data": Array [
            Object {
              "x": "16.67",
            },
          ],
          "label": "ETH",
        },
      ],
    }
  }
  options={
    Object {
      "legend": Object {
        "display": true,
        "labels": Object {
          "fontColor": "rgba(0, 0, 0, 0.87)",
        },
      },
      "maintainAspectRatio": false,
      "responsive": true,
      "scales": Object {
        "xAxes": Array [
          Object {
            "gridLines": Object {
              "color": "rgba(0, 0, 0, 0.54)",
              "display": false,
            },
            "stacked": true,
            "ticks": Object {
              "fontColor": "rgba(0, 0, 0, 0.87)",
              "max": 100,
              "min": 0,
            },
          },
        ],
        "yAxes": Array [
          Object {
            "gridLines": Object {
              "color": "rgba(0, 0, 0, 0.54)",
            },
            "stacked": true,
          },
        ],
      },
      "tooltips": Object {
        "callbacks": Object {
          "label": [Function],
        },
        "mode": "x",
      },
    }
  }
/>
`;
